File: cpp/2328.cpp
#include <map>
#include <utility>
#include <vector>

using namespace std;

class Solution {
private:
  const int MOD = 1e9 + 7;

public:
  map<pair<int, int>, int> memo;
  int helper(vector<vector<int>> &grid, int r = 0, int c = 0, int prev = 0) {
    pair<int, int> key = {r, c};
    if (r < 0 || c < 0 || r == grid.size() || c == grid[0].size())
      return 0;
    else if (grid[r][c] <= prev) // not strictly increase
      return 0;
    else if (memo.find(key) != memo.end())
      return memo[key] % MOD;

    memo[key] = ((((1 + helper(grid, r, c + 1, grid[r][c])) % MOD +
                   helper(grid, r, c - 1, grid[r][c])) %
                      MOD +
                  helper(grid, r + 1, c, grid[r][c])) %
                     MOD +
                 helper(grid, r - 1, c, grid[r][c])) %
                MOD;

    return memo[key];
  }
  int countPaths(vector<vector<int>> &grid) {
    int ret = 0;
    for (int i = 0; i < grid.size(); i++)
      for (int j = 0; j < grid[0].size(); j++)
        ret = (ret + helper(grid, i, j)) % MOD;
    return ret;
  }
};

File: python/2328.py
from typing import List


class Solution:
    def __init__(self):
        self.__memo = dict()
        self.__mod = 1e9+7

    def __helper(self,
                 grid: List[List[int]],
                 r: int, c: int,
                 prev: int = 0) -> int:
        key = (r, c)
        if r < 0 or c < 0 or r == len(grid) or c == len(grid[0]):
            return 0
        elif grid[r][c] <= prev:
            return 0
        elif key in self.__memo:
            return self.__memo[key]

        self.__memo[key] = \
            ((((1 +
             self.__helper(grid, r, c+1, grid[r][c])) % self.__mod +
             self.__helper(grid, r, c-1, grid[r][c])) % self.__mod +
             self.__helper(grid, r+1, c, grid[r][c])) % self.__mod +
             self.__helper(grid, r-1, c, grid[r][c])) % self.__mod

        return self.__memo[key]

    def countPaths(self, grid: List[List[int]]) -> int:
        sum = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                sum = (sum + self.__helper(grid, i, j)) % self.__mod
        return int(sum)
